[Kubelet] Improving QOS in kubelet by introducing QoS level Cgroups - `--cgroups-per-qos`

dubstack opened the pull request:
This PR is tied to this upstream issue #27204 
Please note that only the last commit is unique to this PR. The first two commits are from previous PR's.

It introduces a new flag in the Kubelet which can be used to specify if the user wants to use the QoS cgroup hierarchy. 

cc @kubernetes/sig-node

vishh commented on pkg/kubelet/cm/types.go:
```diff
@@ -49,3 +49,16 @@ type CgroupManager interface {
+}
+
+// QOSContainersInfo hold the names of containers per qos
+type QOSContainersInfo struct {
```
Question: Why is this required?

dubstack replied:
I thought that this would be good have. Once the top level QOS containers are created they won't change. So we can store there names in the cm and it can be used for getting the absolute names of the pod containers.

vishh commented on pkg/kubelet/cm/helpers.go:
```diff
@@ -65,6 +65,7 @@ func getLibcontainerCgroupManager(cgroupConfig *CgroupConfig) (*cgroupfs.Manager
 	// Extract the cgroup resource parameters
 	resourceConfig := cgroupConfig.ResourceParameters
 	resources := &libcontainerconfigs.Resources{}
+	resources.AllowAllDevices = true
```
nit: Move this line into your second commit - libcontainer cgroup manager implementation.

dubstack replied:
Actually I don't think we need this line. This line got in by mistake when I was testing the code.
I will remove this line

dubstack replied:
We needed this line so have made changes in both PR's

vishh commented on pkg/kubelet/cm/container_manager_linux.go:
```diff
@@ -240,6 +282,18 @@ func (cm *containerManagerImpl) setupNode() error {
+		qosConfig := &QOSConfig{
+			RootContainerName: cm.NodeConfig.CgroupRoot,
+		}
+		qosContainersInfo, err := InitQOS(qosConfig)
```
If you are passing only one parameter why create a config object to pass that? Why not pass `RootContainerName` as an argument directly?

dubstack replied:
Okay. Yes i make the change.

vishh commented on pkg/kubelet/cm/container_manager_linux.go:
```diff
@@ -190,6 +193,45 @@ const (
+// We currently create containers for the Burstable and Best Effort containers
+// All guaranteed pods are nested under the RootContainer by default
+// Init is called only once during kubelet bootstrapping.
+func InitQOS(qosConfig *QOSConfig) (QOSContainersInfo, error) {
```
nit: If you accept pointer arguments, you should generally validate that it is not nil.

dubstack replied:
Ack! Thanks!

vishh commented on pkg/kubelet/cm/container_manager_linux.go:
```diff
@@ -190,6 +193,45 @@ const (
+	// The rootContainer under which all qos containers are brought up is configurable
+	// and can be specified through the --cgroup-root flag.
+	// We default to the system root / in case no root is specified
+	rootContainerName := defaultContainerRoot
```
We should not default here. Instead, we should default via flags. AFAIK `--cgroup-root` is already defaulted to `/`.

dubstack replied:
I believe currently cgroup root defaults to "". and not "/" hence the default.

vishh replied:
In that case, let's add the default via the flag, to make it explicit to
the user.

On Thu, Jun 23, 2016 at 4:28 PM, Buddha Prakash <notifications@github.com>
wrote:

> In pkg/kubelet/cm/container_manager_linux.go
> <https://github.com/kubernetes/kubernetes/pull/27853#discussion_r68332856>
> :
>
> > @@ -190,6 +193,45 @@ const (
> >  	KernelTunableModify KernelTunableBehavior = "modify"
> >  )
> >
> > +// Init creates the top level qos cgroup containers
> > +// We currently create containers for the Burstable and Best Effort containers
> > +// All guaranteed pods are nested under the RootContainer by default
> > +// Init is called only once during kubelet bootstrapping.
> > +func InitQOS(qosConfig *QOSConfig) (QOSContainersInfo, error) {
> > +	cm := NewCgroupManagerImpl()
> > +	// The rootContainer under which all qos containers are brought up is configurable
> > +	// and can be specified through the --cgroup-root flag.
> > +	// We default to the system root / in case no root is specified
> > +	rootContainerName := defaultContainerRoot
>
> I believe currently cgroup root defaults to "". and not "/" hence the
> default.
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/27853/files/773b7e7144cb41cbb65d45b4e902e3888636af17..274b013b331ab329a043a235d717b1d3d578811b#r68332856>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AGvIKNYuqpRpf78z7IkYcMEmh3aUecJMks5qOxaCgaJpZM4I7o6F>
> .
>

dubstack replied:
Cgroup-root default "" has been used at a lot places in the codebase. I thought I will once again confirm with you before I make it default to "/".

vishh commented:
I added a few comments. Ping me back once those comments are addressed

dubstack commented:
@vishh I added the e2e test, but I am not sure how to access the host node cgroup fs information from inside the container. Is there a hack to achieve that? I have addressed all your other comments except changing the default of cgroup-root to "/".

vishh commented:
You can mount in `/sys/fs/cgroup` using a HostPath volume

On Fri, Jun 24, 2016 at 6:07 AM, Buddha Prakash <notifications@github.com>
wrote:

> @vishh <https://github.com/vishh> I added the e2e test, but I am not sure
> how to access the host node cgroup fs information from inside the
> container. Is there a hack to achieve that? I have addressed all your other
> comments except changing the default of cgroup-root to "/".
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/27853#issuecomment-228340379>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AGvIKOxjreoIiR_grw3xEBb9KRNspFWDks5qO9apgaJpZM4I7o6F>
> .
>

derekwaynecarr commented on pkg/apis/componentconfig/types.go:
```diff
@@ -252,6 +252,10 @@ type KubeletConfiguration struct {
+	// Enable Qos based Cgroup hieracrchy: top level cgroups for Qos Classes
+	// And all pods are broguht up under a top level Qos cgroup
+	// based on the Qos class they belong to.
+	EnableQosCgroups bool `json:"enablePodCgroups,omitempty"`
```
see mismatch between var name and json.

dubstack replied:
Ack! Thanks!

derekwaynecarr commented on pkg/apis/componentconfig/types.go:
```diff
@@ -252,6 +252,10 @@ type KubeletConfiguration struct {
 	// KubeletCgroups is the absolute name of cgroups to isolate the kubelet in.
 	KubeletCgroups string `json:"kubeletCgroups,omitempty"`
+	// Enable Qos based Cgroup hieracrchy: top level cgroups for Qos Classes
+	// And all pods are broguht up under a top level Qos cgroup
```
brought

dubstack replied:
Ack! Thanks!

derekwaynecarr commented on pkg/kubelet/cm/container_manager_linux.go:
```diff
@@ -190,6 +193,45 @@ const (
+	// Create containers for both qos classes
+	for _, qosClass := range qosClasses {
+		//get the containers absolute name
+		absoluteContainerName := rootContainerName + string(qosClass)
```
please use path.Join

derekwaynecarr replied:
typically, i would expect the qosClass to have also have a toLower so Guaranteed is just guaranteed

dubstack replied:
@derekwaynecarr  I will be making some changes and rebasing. I will ping you once I am done with those changes.

dubstack replied:
And lets have the toLower() method in a seperate PR?

dubstack committed:
Inject top level QoS cgroup creation in the Kubelet

dubstack commented:
@derekwaynecarr @Random-Liu  @vishh  PTAL. We decided to follow the policy that if a user wants to enable the qos Cgroup hierarchy he would have to specify a valid cgroup Root. This is because for now we dont want to assign default values to CgroupRoot in a way which is hidden from the user. 

Note that the GCE Node tests would complain as tests run on the coreos image which uses systemd as the cgroup driver will fail.

We can consider other policies. We can discuss them over here.

dubstack commented:
@k8s-bot test this issue: #IGNORE

Random-Liu commented:
@dubstack Will review today.

Random-Liu commented on cmd/kubelet/app/options/options.go:
```diff
@@ -229,6 +230,7 @@ func (s *KubeletServer) AddFlags(fs *pflag.FlagSet) {
 	fs.MarkDeprecated("system-container", "Use --system-cgroups instead. Will be removed in a future version.")
 	fs.StringVar(&s.SystemCgroups, "system-cgroups", s.SystemCgroups, "Optional absolute name of cgroups in which to place all non-kernel processes that are not already inside a cgroup under `/`. Empty for no container. Rolling back the flag requires a reboot. (Default: \"\").")
 
+	fs.BoolVar(&s.CgroupPerQOS, "cgroup-per-qos", s.CgroupPerQOS, "Whether the Kubelet should have the QoS cgroup hierarchy, if true top level qos and pod cgroups are created.")
```
nits: explicitly explain what the default value is.

vishh replied:
nit: rename s/cgroup/cgroups here and in the rest of this PR.

Random-Liu commented on cmd/kubelet/app/server.go:
```diff
@@ -358,12 +359,16 @@ func run(s *options.KubeletServer, kcfg *KubeletConfig) (err error) {
 		}
-
+		if kcfg.CgroupPerQOS && kcfg.CgroupRoot == "" {
+			return fmt.Errorf("invalid configuration: CgroupPerQOS was specified and cgroup root was not specified. To enable the QoS cgroup hierarchy you need tp specify a valid cgroupRoot")
```
s/tp/to

Random-Liu commented on pkg/kubelet/cm/container_manager_linux.go:
```diff
@@ -97,7 +98,10 @@ type containerManagerImpl struct {
 	systemContainers []*systemContainer
+	qosContainers    QOSContainersInfo
 	periodicTasks    []func()
+	// holds all the mounted cgroup subsytems
```
s/subsytems/subsystems

Random-Liu commented on pkg/kubelet/cm/container_manager_linux.go:
```diff
@@ -190,6 +199,38 @@ const (
+
+	// Create containers for both qos classes
+	for _, qosClass := range qosClasses {
+		//get the containers absolute name
```
space missing.

Random-Liu commented on pkg/kubelet/kubelet.go:
```diff
@@ -727,6 +729,9 @@ type Kubelet struct {
 	// Monitor resource usage
 	resourceAnalyzer stats.ResourceAnalyzer
 
+	// Whether or not we should have the QOS cgroup Hierarchy for resource Management
```
s/Management/management

Random-Liu commented on test/e2e_node/cgroup_manager_test.go:
```diff
@@ -0,0 +1,70 @@
+/*
```
Why not just check the local directory? This is node e2e test right?

Random-Liu replied:
Offline discussed with @dubstack, in the future we'll also check whether the container cgroup is set correctly. Anyway we have to start a container, then we just check the cgroup directory in the container.

vishh replied:
One more thing is that, running the test in pod makes the test portable across cluster and node e2es.
WDYT @Random-Liu ?

Random-Liu replied:
@vishh Yeah, SGTM. :)

vishh commented on pkg/kubelet/cm/container_manager_linux.go:
```diff
@@ -190,6 +199,38 @@ const (
+		// containerConfig object stores the cgroup specifications
+		containerConfig := &CgroupConfig{
+			Name:               absoluteContainerName,
+			ResourceParameters: &ResourceConfig{},
```
Don't we have to allow access to all devices here or is that the default setup for `Create`?

dubstack replied:
Yes by default create sets allowalldevices : true. Should I change that behaviour?

vishh commented on pkg/kubelet/cm/container_manager_linux.go:
```diff
@@ -190,6 +199,38 @@ const (
+	// Create containers for both qos classes
+	for _, qosClass := range qosClasses {
+		//get the containers absolute name
+		absoluteContainerName := path.Join(rootContainer, string(qosClass))
```
We should ensure that `rootContainer` exists in reality. What if I pass a random path like `foo13#$$` for `--cgroup-root`?

dubstack replied:
Yes I thought it would be a good idea to have that check over here: https://github.com/kubernetes/kubernetes/blob/master/cmd/kubelet/app/server.go#L360

WDYT??

vishh commented on test/e2e_node/cgroup_manager_test.go:
```diff
@@ -0,0 +1,70 @@
+					},
+					Spec: api.PodSpec{
+						// Force the Pod to schedule to the node without a scheduler running
+						NodeName: *nodeName,
```
nit: This is no longer needed. Use the [pod helper](https://github.com/kubernetes/kubernetes/blob/master/test/e2e/framework/pods.go#L36) methods instead.

dubstack replied:
Ack!

dubstack replied:
I use TestContainerOutput() which creates the pod. So I can't use the pod helper for creating the pods in this case.

dubstack replied:
re: above, I spoke to Tim and got to know that he also added one more function MungePodSpec() which adds the pod
So I will check if that can be used.

vishh commented:
Added a couple of comments @dubstack. Ping me back once you have addressed all the existing comments.

dubstack commented:
@vishh Addressed all comments. PTAL

dubstack commented:
@k8s-bot e2e test this ##26760

dubstack commented:
@k8s-bot e2e test this #IGNORE

dubstack commented:
@vishh Can we get this merged?

ncdc commented on pkg/kubelet/cm/container_manager_linux.go:
```diff
@@ -161,10 +165,24 @@ func validateSystemRequirements(mountUtil mount.Interface) (features, error) {
+	}
+	subsystems, err := getCgroupSubsystems()
+	if err != nil {
+		return &containerManagerImpl{}, fmt.Errorf("Failed to get mounted subsystems: %v", err)
```
Make "Failed" lowercase

ncdc commented on pkg/kubelet/cm/container_manager_linux.go:
```diff
@@ -161,10 +165,24 @@ func validateSystemRequirements(mountUtil mount.Interface) (features, error) {
+			return &containerManagerImpl{}, fmt.Errorf("invalid configuration: CgroupRoot doesn't exist : %v", err)
+		}
+	}
+	if nodeConfig.CgroupsPerQOS && nodeConfig.CgroupRoot == "" {
```
Shouldn't this check come before the previous one (`os.Stat`)? Also, why not just do:

```go
if nodeConfig.CgroupsPerQOS {
  if nodeConfig.CgroupRoot == "" {
    // return error
  }
  if _, err := os.Stat(nodeConfig.CgroupRoot); err != nil {
    // return error
  }
}
```

derekwaynecarr replied:
this check should come before the check to see if the file directory exists above

dubstack replied:
Ack! Thanks!

ncdc commented on pkg/kubelet/cm/container_manager_linux.go:
```diff
@@ -190,6 +208,38 @@ const (
+			ResourceParameters: &ResourceConfig{},
+		}
+		if err := cm.Create(containerConfig); err != nil {
+			return QOSContainersInfo{}, fmt.Errorf("Failed to create top level %v QOS cgroup : %v", qosClass, err)
```
lowercase failed

derekwaynecarr commented on cmd/kubelet/app/options/options.go:
```diff
@@ -232,6 +233,7 @@ func (s *KubeletServer) AddFlags(fs *pflag.FlagSet) {
 	fs.MarkDeprecated("system-container", "Use --system-cgroups instead. Will be removed in a future version.")
 	fs.StringVar(&s.SystemCgroups, "system-cgroups", s.SystemCgroups, "Optional absolute name of cgroups in which to place all non-kernel processes that are not already inside a cgroup under `/`. Empty for no container. Rolling back the flag requires a reboot. (Default: \"\").")
 
+	fs.BoolVar(&s.CgroupsPerQOS, "cgroups-per-qos", s.CgroupsPerQOS, "Whether the Kubelet should have the QoS cgroup hierarchy, if true top level qos and pod cgroups are created. Default: false")
```
use QoS and qos consistently in this help text.

derekwaynecarr replied:
isnt default: false redundant since the default is rendered in s.CgroupsPerQOS

ncdc commented on pkg/kubelet/cm/container_manager_linux.go:
```diff
@@ -190,6 +208,38 @@ const (
+		}
+	}
+	// Store the top level qos container names
+	qosContainersInfo := QOSContainersInfo{}
```
I usually would do

```go
qosContainersInfo := QOSContainersInfo{
  Guaranteed: rootContainer,
  // ...
}
```

ncdc commented on pkg/kubelet/cm/container_manager_linux.go:
```diff
@@ -240,6 +290,15 @@ func (cm *containerManagerImpl) setupNode() error {
+	if cm.NodeConfig.CgroupsPerQOS {
+		qosContainersInfo, err := InitQOS(cm.NodeConfig.CgroupRoot, cm.subsystems)
+		if err != nil {
+			return fmt.Errorf("Failed to initialise top level QOS containers: %v", err)
```
lowercase failed

derekwaynecarr replied:
errors should start with lowercase.

also typo on initialize

derekwaynecarr commented on pkg/apis/componentconfig/types.go:
```diff
@@ -252,13 +252,17 @@ type KubeletConfiguration struct {
 	CloudConfigFile string `json:"cloudConfigFile,omitempty"`
 	// KubeletCgroups is the absolute name of cgroups to isolate the kubelet in.
 	KubeletCgroups string `json:"kubeletCgroups,omitempty"`
+	// Enable QoS based Cgroup hieracrchy: top level cgroups for QoS Classes
```
typo: hierarchy

derekwaynecarr commented on pkg/apis/componentconfig/types.go:
```diff
@@ -252,13 +252,17 @@ type KubeletConfiguration struct {
 	// for no container. Rolling back the flag requires a reboot.
 	SystemCgroups string `json:"systemContainer,omitempty"`
-	// cgroupRoot is the root cgroup to use for pods. This is handled by the
+	// CgroupRoot is the root cgroup to use for pods. This is handled by the
```
we dont really need to lead the help text with the field name anymore.

"This is the root cgroup to use for pods.  If CgroupsPerQOS is enabled, this is the root of the QoS cgroup hierarchy"

derekwaynecarr commented on pkg/kubelet/cm/container_manager_linux.go:
```diff
@@ -161,10 +165,24 @@ func validateSystemRequirements(mountUtil mount.Interface) (features, error) {
+		}
+	}
+	if nodeConfig.CgroupsPerQOS && nodeConfig.CgroupRoot == "" {
+		return &containerManagerImpl{}, fmt.Errorf("invalid configuration: CgroupsPerQOS was specified and cgroup root was not specified. To enable the QoS cgroup hierarchy you need to specify a valid cgroupRoot")
```
nit: be consistent on cgroup root and cgroupRoot

derekwaynecarr commented on pkg/kubelet/cm/container_manager_linux.go:
```diff
@@ -190,6 +208,38 @@ const (
 )
 
+// InitQOS creates the top level qos cgroup containers
+// We currently create containers for the Burstable and Best Effort containers
```
We currently create containers for the Burstable and BestEffort classes or pods (not sure what you were intending)

derekwaynecarr commented on pkg/kubelet/cm/container_manager_linux.go:
```diff
@@ -190,6 +208,38 @@ const (
+// We currently create containers for the Burstable and Best Effort containers
+// All guaranteed pods are nested under the RootContainer by default
+// InitQOS is called only once during kubelet bootstrapping.
+func InitQOS(rootContainer string, subsystems *cgroupSubsystems) (QOSContainersInfo, error) {
```
Add a // TODO add support for cgroupRoot?

derekwaynecarr commented on pkg/kubelet/cm/container_manager_linux.go:
```diff
@@ -190,6 +208,38 @@ const (
+			Name:               absoluteContainerName,
+			ResourceParameters: &ResourceConfig{},
+		}
+		if err := cm.Create(containerConfig); err != nil {
```
maybe a note or TODO that this is not yet supported on systemd ;-)

derekwaynecarr commented on pkg/kubelet/cm/container_manager_linux.go:
```diff
@@ -240,6 +290,15 @@ func (cm *containerManagerImpl) setupNode() error {
 
+	// Setup top level qos containers only if CgroupsPerQOS flag is specified as true
+	if cm.NodeConfig.CgroupsPerQOS {
+		qosContainersInfo, err := InitQOS(cm.NodeConfig.CgroupRoot, cm.subsystems)
```
who ensures they are not deleted after the fact?

derekwaynecarr replied:
for your own future sanity, i think you will want to ensure that these are never deleted by a rogue init system.  its fine to handle that in a future todo if @vishh agrees.

vishh replied:
We talked about ensuring that qos cgroups continues to exist whenever a pod
level cgroup is being created. IIRC, @dubstack is going to add that
functionality in a future PR.

On Fri, Jul 8, 2016 at 2:38 PM, Derek Carr <notifications@github.com> wrote:

> In pkg/kubelet/cm/container_manager_linux.go
> <https://github.com/kubernetes/kubernetes/pull/27853#discussion_r70144806>
> :
>
> > @@ -240,6 +290,15 @@ func (cm *containerManagerImpl) setupNode() error {
> >  		return err
> >  	}
> >
> > +	// Setup top level qos containers only if CgroupsPerQOS flag is specified as true
> > +	if cm.NodeConfig.CgroupsPerQOS {
> > +		qosContainersInfo, err := InitQOS(cm.NodeConfig.CgroupRoot, cm.subsystems)
>
> for your own future sanity, i think you will want to ensure that these are
> never deleted by a rogue init system. its fine to handle that in a future
> todo if @vishh <https://github.com/vishh> agrees.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/27853/files/8134b9de63922520d329b09348c13f5cc110507e#r70144806>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AGvIKM8pg3kWyFLjOpd4zq99F5RKmqdpks5qTsNBgaJpZM4I7o6F>
> .
>

derekwaynecarr commented on pkg/kubelet/cm/types.go:
```diff
@@ -54,3 +54,10 @@ type CgroupManager interface {
 }
+
+// QOSContainersInfo hold the names of containers per qos
+type QOSContainersInfo struct {
```
@vishh - we should think about returning summary stats per qos tier in kubelet summary api

vishh replied:
+1

On Fri, Jul 8, 2016 at 2:35 PM, Derek Carr <notifications@github.com> wrote:

> In pkg/kubelet/cm/types.go
> <https://github.com/kubernetes/kubernetes/pull/27853#discussion_r70144463>
> :
>
> > @@ -54,3 +54,10 @@ type CgroupManager interface {
> >  	// Update cgroup configuration.
> >  	Update(*CgroupConfig) error
> >  }
> > +
> > +// QOSContainersInfo hold the names of containers per qos
> > +type QOSContainersInfo struct {
>
> @vishh <https://github.com/vishh> - we should think about returning
> summary stats per qos tier in kubelet summary api
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/27853/files/8134b9de63922520d329b09348c13f5cc110507e#r70144463>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AGvIKLZsWf-mnBjVGJQIJJ8scuzqTv5Dks5qTsKMgaJpZM4I7o6F>
> .
>

derekwaynecarr commented on pkg/kubelet/kubelet.go:
```diff
@@ -731,6 +733,9 @@ type Kubelet struct {
 	// Monitor resource usage
 	resourceAnalyzer stats.ResourceAnalyzer
 
+	// Whether or not we should have the QOS cgroup Hierarchy for resource management
```
toLower(Hierarchy)

derekwaynecarr commented:
few nits that once addressed, this lgtm

euank commented on pkg/kubelet/cm/container_manager_linux.go:
```diff
@@ -161,10 +165,24 @@ func validateSystemRequirements(mountUtil mount.Interface) (features, error) {
+	// Check if Cgroup-root actually exists on the node
+	if nodeConfig.CgroupsPerQOS {
+		if _, err := os.Stat(nodeConfig.CgroupRoot); err != nil {
+			return &containerManagerImpl{}, fmt.Errorf("invalid configuration: CgroupRoot doesn't exist : %v", err)
```
nit, any reason not to return `nil` here and in the other error cases?

dubstack replied:
No I think it should not matter.

dubstack commented:
@euank @vishh @derekwaynecarr @ncdc Thanks a lot for your comments. I have addressed all nits. PTAL

euank commented:
LGTM

dubstack commented:
@vishh @derekwaynecarr can i get an LGTM??

derekwaynecarr commented on cmd/kubelet/app/options/options.go:
```diff
@@ -232,6 +233,7 @@ func (s *KubeletServer) AddFlags(fs *pflag.FlagSet) {
 	fs.MarkDeprecated("system-container", "Use --system-cgroups instead. Will be removed in a future version.")
 	fs.StringVar(&s.SystemCgroups, "system-cgroups", s.SystemCgroups, "Optional absolute name of cgroups in which to place all non-kernel processes that are not already inside a cgroup under `/`. Empty for no container. Rolling back the flag requires a reboot. (Default: \"\").")
 
+	fs.BoolVar(&s.CgroupsPerQOS, "cgroups-per-qos", s.CgroupsPerQOS, "Whether the Kubelet should have the QoS cgroup hierarchy, if true top level QoS and pod cgroups are created. Default: false")
```
Nit: remove the default: false piece

The default is rendered in help based on s.CgroupsPerQOS already.

dubstack replied:
Ack

derekwaynecarr commented on test/e2e_node/e2e_service.go:
```diff
@@ -235,6 +235,8 @@ func (es *e2eService) startKubeletServer() (*killCmd, error) {
 		"--v", LOG_VERBOSITY_LEVEL, "--logtostderr",
 		"--network-plugin=kubenet",
 		"--pod-cidr=10.180.0.0/24", // Assign a fixed CIDR to the node because there is no node controller.
+		"--cgroups-per-qos", "true",
```
This will break node e2e runs on RHEL which we have worked hard to get passing.  Can we make testing of this controlled via some flag or other input into node e2e so systemd based environments don't fail after this PR merges?

dubstack replied:
Oh! I completely overlooked this. I agree with you but I am not sure how we can achieve that. So correct me if I am wrong. To achieve this
1. the e2eservice needs to detect if its running on a systemd system.
2. if its running on a systemd system then kubelet is started without the cgroups-per-qos flag
3. We set some flag so that the cgroup hierarchy tests do not run when this flag is set.
I think we can use this https://github.com/dubstack/kubernetes/blob/0a7b6cd2c88f561f4d8c3fb9a095d522ec33dc6d/test/e2e_node/e2e_service.go#L211
for first point
But I am not sure how to handle point 3.

ncdc replied:
I don't think you need to detect systemd, do you? Can't you just check an
env var, and we'd set that manually prior to running the test? You could
also skip the test based on the value of the env var.

On Saturday, July 9, 2016, Buddha Prakash <notifications@github.com> wrote:

> In test/e2e_node/e2e_service.go
> <https://github.com/kubernetes/kubernetes/pull/27853#discussion_r70162085>
> :
>
> > @@ -235,6 +235,8 @@ func (es *e2eService) startKubeletServer() (*killCmd, error) {
> >  		"--v", LOG_VERBOSITY_LEVEL, "--logtostderr",
> >  		"--network-plugin=kubenet",
> >  		"--pod-cidr=10.180.0.0/24", // Assign a fixed CIDR to the node because there is no node controller.
> > +		"--cgroups-per-qos", "true",
>
> Oh! I completely overlooked this. I agree with you but I am not sure how
> we can achieve that. So correct me if I am wrong. To achieve this
> 1. the e2eservice needs to detect if its running on a systemd system.
> 2. if its running on a systemd system then kubelet is started without the
> cgroups-per-qos flag
> 3. We set some flag so that the cgroup hierarchy tests do not run when
> this flag is set.
> I think we can use this
> https://github.com/dubstack/kubernetes/blob/0a7b6cd2c88f561f4d8c3fb9a095d522ec33dc6d/test/e2e_node/e2e_service.go#L211
> for first point
> But I am not sure how to handle point 3.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/27853/files/0a7b6cd2c88f561f4d8c3fb9a095d522ec33dc6d#r70162085>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AAABYhAFx2zigzELSt_gjTKdwrzfnm-zks5qTx3fgaJpZM4I7o6F>
> .
>

vishh replied:
+1 for adding a *temporary* flag to node e2e, to disable this feature on systemd nodes.

dubstack replied:
Sounds Good. 👍

derekwaynecarr commented:
I have one remaining nit and a request to not break node e2e runs we run on
RHEL by having a flag or separate conformance check that you guys enable on
your environments.

On Friday, July 8, 2016, Buddha Prakash <notifications@github.com> wrote:

> @vishh <https://github.com/vishh> @derekwaynecarr
> <https://github.com/derekwaynecarr> can i get an LGTM??
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/27853#issuecomment-231509540>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AF8dbGQwxLAth9RMOl-RI8ULDNzSB8hZks5qTwX-gaJpZM4I7o6F>
> .
>

dubstack commented:
@derekwaynecarr @ncdc I added a cgroups-per-qos flag which can be set to false if a user wants to disable the qos cgroup hierarchy creation and testing. 
@vishh Gingko doen't initialize the testContext flags until one of the spec is run, So i am not sure where to have the check for running the QoS cgroups creation tests. The way I have done it currently can be a bit misleading.

vishh commented on test/e2e_node/cgroup_manager_test.go:
```diff
@@ -0,0 +1,71 @@
+								{
+									Image:   "gcr.io/google_containers/busybox:1.24",
+									Name:    podName,
+									Command: []string{"sh", "-c", "if [ -d /tmp/memory/Burstable ]; then echo Found; else echo Failed; fi"},
```
Check for `Besteffort` as well. I'd recommend `exit 1` instead of logging.

vishh commented on test/e2e_node/e2e_remote.go:
```diff
@@ -211,12 +211,18 @@ func RunRemote(archive string, host string, cleanup bool, junitFileNumber int, s
-		fmt.Sprintf("cd %s", tmp),
-		fmt.Sprintf("timeout -k 30s %ds ./ginkgo %s ./e2e_node.test -- --logtostderr --v 2 --build-services=false --stop-services=%t --node-name=%s --report-dir=%s/results --junit-file-number=%d", *testTimeoutSeconds, *ginkgoFlags, cleanup, host, tmp, junitFileNumber),
-	)
+	if cgroupsPerQOS {
```
nit: unify these two statements to avoid duplication.

dubstack commented:
@derekwaynecarr @ncdc PTAL, I added a flag which will skip these tests by default. 

@vishh Please do a final review.

vishh commented on pkg/apis/componentconfig/deep_copy_generated.go:
```diff
@@ -1,300 +0,0 @@
-// +build !ignore_autogenerated
```
Why is this file being deleted?

dubstack replied:
I think a PR went in which included the auto code generation in the build system.

vishh commented on test/e2e_node/e2e_service.go:
```diff
@@ -234,10 +239,15 @@ func (es *e2eService) startKubeletServer() (*killCmd, error) {
 		"--config", es.kubeletStaticPodDir,
 		"--file-check-frequency", "10s", // Check file frequently so tests won't wait too long
 		"--v", LOG_VERBOSITY_LEVEL, "--logtostderr",
+		"--network-plugin=kubenet",
```
why are you adding this?

dubstack replied:
This was added on by some other PR aswell. I am not sure why its showing up as green.

dubstack replied:
Maybe because the first commit doesen't have it

dubstack commented:
Adding LGTM after Rebase.
